packages/bff/src/services/agreementService.ts

```diff
@@ -57,3 +149,279 @@ export const getLatestAgreement = async (
+    certifiedAttributes: agreementCertifiedAttrs,
+    declaredAttributes: agreementDeclaredAttrs,
+    suspendedByConsumer: agreement.suspendedByConsumer,
+    suspendedByProducer: agreement.suspendedByProducer,
+    isContractPresent: agreement.contract !== undefined,
+    consumerDocuments: agreement.consumerDocuments,
+    createdAt: agreement.createdAt,
+    updatedAt: agreement.updatedAt,
+    suspendedAt: agreement.suspendedAt,
+  };
+}
+
+function descriptorAttributesIds(
+  descriptor: catalogApi.EServiceDescriptor
+): string[] {
+  const { verified, declared, certified } = descriptor.attributes;
+  const allAttributes = [
+    ...verified.flat(),
+    ...declared.flat(),
+    ...certified.flat(),
+  ];
+  return allAttributes.map((attr) => attr.id);
+}
+
+function tenantAttributesIds(tenant: tenantApi.Tenant): string[] {
+  const verifiedIds = tenant.attributes.map((attr) => attr.verified?.id);
+  const certifiedIds = tenant.attributes.map((attr) => attr.certified?.id);
+  const declaredIds = tenant.attributes.map((attr) => attr.declared?.id);
+
+  return [...verifiedIds, ...certifiedIds, ...declaredIds].filter(
+    (x): x is string => x !== undefined
+  );
+}
+
+async function parallelGet(
+  agreement: agreementApi.Agreement,
+  { tenantProcessClient, catalogProcessClient }: PagoPAInteropBeClients,
+  { headers }: WithLogger<BffAppContext>
+): Promise<{
+  consumer: tenantApi.Tenant;
+  producer: tenantApi.Tenant;
+  eservice: catalogApi.EService;
+}> {
+  const consumerTask = tenantProcessClient.tenant.getTenant({
+    params: { id: agreement.consumerId },
+    headers,
+  });
+
+  const producerTask = tenantProcessClient.tenant.getTenant({
+    params: { id: agreement.producerId },
+    headers,
+  });
+  const eserviceTask = catalogProcessClient.getEServiceById({
+    params: { eServiceId: agreement.eserviceId },
+    headers,
+  });
+  const [consumer, producer, eservice] = await Promise.all([
+    consumerTask,
+    producerTask,
+    eserviceTask,
+  ]);
+
+  return {
+    consumer,
+    producer,
+    eservice,
+  };
+}
+
+async function getBulkAttributes(
+  ids: string[],
+  attributeProcess: PagoPAInteropBeClients["attributeProcessClient"],
+  { headers }: WithLogger<BffAppContext>
+): Promise<attributeRegistryApi.Attribute[]> {
+  async function getAttributesFrom(
+    offset: number
+  ): Promise<attributeRegistryApi.Attribute[]> {
+    const response = await attributeProcess.getBulkedAttributes(ids, {
+      queries: { offset, limit: 50 },
+      headers,
+    });
+    return response.results;
+  }
+
+  async function aggregate(
+    start: number,
+    attributes: attributeRegistryApi.Attribute[]
+  ): Promise<attributeRegistryApi.Attribute[]> {
+    const attrs = await getAttributesFrom(start);
+    if (attrs.length < 50) {
+      return attributes.concat(attrs);
+    } else {
+      return aggregate(start + 50, attributes.concat(attrs));
+    }
+  }
+
+  return aggregate(0, []);
+}
```

Comment:
again in this method is equals to other one in another [PR](https://github.com/pagopa/interop-be-monorepo/blob/07e43f519a29cdec065110f0b4981660442a95f0/packages/bff/src/services/catalogService.ts#L116), we should move it in a dedicated file.